include/caffeine/ADT/ClonePointer.h

```diff
@@ -0,0 +1,53 @@
+#include <memory>
+
+namespace caffeine {
+
+// This class essentially wraps a unique pointer. However, instead of
+// copying the same pointer every time this wrapper is copied, the
+// actual object behind the pointer is copied as well.
+//
+// This is desirable because we are then able to have subclasses in a
+// vector when normally that would not be possible.
+//
+// In order to use this class, the template parameter T must implement
+// a clone() method which returns a std::unique_ptr<T>. This method is
+// called on copys and assignments in order to copy the underlying class
+// behind the unique pointer.
+template <class T>
+class clone_ptr {
+  std::unique_ptr<T> pointer;
+
+public:
+  virtual ~clone_ptr() = default;
+
+  clone_ptr(std::unique_ptr<T>&& p) : pointer(std::move(p)) {}
+
+  clone_ptr(const clone_ptr<T>& other) : pointer(other.pointer->clone()) {}
+
+  clone_ptr& operator=(const clone_ptr<T> other) {
+    pointer = other.pointer->clone();
+    return *this;
+  }
+
+  T* operator->() const {
+    return pointer.get();
+  }
+
+  T* get() const {
+    return pointer.get();
+  }
```

Comment:
Add some other operators that `std::unique_ptr` has.
```suggestion
  }
  
  bool operator!() const noexcept {
    return !get();
  }
  
  operator bool() const noexcept {
    return !!get();
  }
  
  bool operator==(const clone_ptr<T>& p) const noexcept {
    return pointer == p.pointer;
  }
  
  bool operator!=(const clone_ptr<T>& p) const noexecept {
    return !(*this == p);
  }
```